### ISSUE TITLE
Interrupt Conversation with Callback Query Handler

### ISSUE THREAD

**Issue by aon** [python-telegram-bot/python-telegram-bot#2079]
### Issue I am facing
I can't interrupt my conversation with a callback query handler.

The thing is I've got a keyboard that runs after /start and one of the options is to start the 'conversation' below. When the conversation starts, it calls to a submenu, that has a few options and then a "Back" button. That "Back" button is set to go back to the main menu, so I call the main menu again (this time differently, given I have to answer the query and all that).

I've set the conv_handler in group 1 and the main menu in group 0. Nevertheless, it's not stopping the conversation after it goes back. By that I mean, it goes to STATE1 and takes the input of the conversation to do something, which I don't want. Is there something wrong?

By the way, I've tried giving each of this states a pattern and that actually works, the thing is this "patterns" would change each time given the options change and besides that it should interrupt my conv if I run a handler with a lower group right?

### Related part of my code
```python

def main(update, context):
    query = update.callback_query
    query.answer()
    query.edit_message_text(
        text="Hi! Choose an option:",
        reply_markup=keyboard
    )
dispatcher.add_handler(CallbackQueryHandler(main, pattern='main'), group=0)

conv_handler = ConversationHandler(
	entry_points=[CallbackQueryHandler(entry, pattern='conversation')],
	states={
		STATE1: [CallbackQueryHandler(state1)],
                STATE2: [CallbackQueryHandler(state2)]
	},
	fallbacks=[CommandHandler('cancel',cancel)]
)
dispatcher.add_handler(conv_handler, group=1)
```

Thanks, I hope I was clear enough.

**Comment by Bibo-Joshi** [comment]
Hi.
If you want to stop your conversation, you will need to return `ConversationHandler.END` from your callback.
That your conversation will try to handle CallbackQueries meant for the main menu is due to the fact, that one handler _per group_ is executed (see the docs of `Dispatcher.add_handler`).
I don't exactly understany what you mean by "changing patterns", but you might consider using regex patterns instead of plain string patterns, if that helps you.
Finally, there are several issues on interruption of conversations, that might help you. Just search them through the search bar on the issues page ;)
Hope that helps.

**Comment by aon** [comment]
Hi, thanks for the quick answer! I checked all the other issues about interrupting conversation. I understood that when a lower group handler is called, then no other handler is called, maybe I'm wrong about this?

Even if I returned `ConversationHandler.END` in the "Back" callback handler, it's still calling the next state in my conversation handler, i.e. not ending the conversation. I'll leave below a more complete version of my code:

```python
# Main menu
def main_keyboard():
    return InlineKeyboardMarkup([
        InlineKeyboardButton("Conversation", callback_data=conversation),
        InlineKeyboardButton("Option 2", callback_data=opt2)
    ])

def main(update, context):
    query = update.callback_query
    query.answer()
    query.edit_message_text(
        text="Hi! Choose an option:",
        reply_markup=main_keyboard()
    )
dispatcher.add_handler(CallbackQueryHandler(main, pattern='main'), group=0)

# Conversation
def entry_keyboard():
    return InlineKeyboardMarkup([
        InlineKeyboardButton("Something else", callback_data=smth_else),
        InlineKeyboardButton("Back", callback_data=main)
    ])

def entry(update, context):
    query = update.callback_query
    query.answer()
    query.edit_message_text(
        text="Alright, choose something else:",
        reply_markup=entry_keyboard()
    )
    return STATE1

def state1(update, context):
    query = update.callback_query
    query.answer()
    context.bot.send_message(
        chat_id=update.effective_chat.id,
        text="You selected: " + query.data
    )
    return ConversationHandler.END

conv_handler = ConversationHandler(
	entry_points=[CallbackQueryHandler(entry, pattern='conversation')],
	states={
		STATE1: [CallbackQueryHandler(state1)],
    },
	fallbacks=[CommandHandler('cancel',cancel)]
)
dispatcher.add_handler(conv_handler, group=1)
```

The functionality I want is when Back is selected, the conversation should end and I should get the previous menu. What I get now is I get the previous menu AND the text "You selected: Something else". Even if in `main()` I return `ConversationHandler.END` it still gets handled by the callback query inside the conversation.

Thank you for your time.

**Comment by Bibo-Joshi** [comment]
> I understood that when a lower group handler is called, then no other handler is called, maybe I'm wrong about this?

You are. Please read the docs of `Dispatcher.add_handler()` again carefully

> The functionality I want is when Back is selected, the conversation should end and I should get the previous menu. What I get now is I get the previous menu AND the text "You selected: Something else".

That's because your `state1` sends that message before returning `ConversationHandler.END`. you should probably change `state1` to do something like

```python
if query.data == 'main':
    update.effective_message.edit_text(…, reply_markup=main_keyboard())
else:
   ....
```

or even use different `CallbackQueryHandler` s for that.

> Even if in `main()` I return `ConversationHandler.END` it still gets handled by the callback query inside the conversation.

`main()` doesn't need to return anything. Just the callbacks in your conversation.

**Comment by aon** [comment]
> You are. Please read the docs of Dispatcher.add_handler() again carefully
Will do, there are some parts I still don't get but will do.

> That's because your state1 sends that message before returning ConversationHandler.END. you should probably change state1 to do something like
That works but it will still call the main CallbackQuery handler. I should put them both in the same group and add the conversation handler first right?

Another way maybe, would be to do
```python
if query.data == 'main':
    return ConversationHandler.END
else:
   ....
```
It will still get called twice but it'll end the conversation.

> or even use different CallbackQueryHandler s for that.
What do you mean by that?

Also I noticed, when I change my STATE1 callback to:
```python
STATE1: [CallbackQueryHandler(state1, pattern='smth_else')]
```
it works as intended, meaning it ends my conversation and goes back to the main menu.

I really searched everywhere, issues, stackoverflow and the docs, and I'm sorry if I missed anything or didn't understand it correctly. Thanks for your patience.

**Comment by Bibo-Joshi** [comment]
>>or even use different CallbackQueryHandler s for that.

> What do you mean by that?

I mean using two `CallbackQueryHandler` s with different `pattern` instead of putting an `if else` in one handler

> when […] it works as intended, meaning it ends my conversation and goes back to the main menu.

Because the CBH won't catch the callbackquery with `data == 'main'` anymore and only your `CallbackQueryHandler(main, pattern='main')` will be called. so your conversation is not really ended, it just doesn't handle the update.

> That works but it will still call the main CallbackQuery handler. I should put them both in the same group and add the conversation handler first right?
 
I can't tell you what you should or shouldn't do. It depends on the use case and how you want the user experience and the code to be designed. What I usually do is have the conversation start on a `CommandHandler` that shows the menu and go on with `CallbackQueryHandler` s after that. Because you detached the displaying of the main menu from the conversation you need to do things differently. There are multiple feasible ways, with those handlers in the same group or not and there is no clear "right" way. If you've understood the takeaway message of `add_handler`, i.e.

> TL;DR: Order and priority counts. 0 or 1 handlers per group will be used. End handling of update with telegram.ext.DispatcherHandlerStop.

I'm sure you can figure out a way to achieve what you want to do. And by what I read I guess your on a good path ;)
I'll go ahead close this issue. For future questions, please be invited to our telegram group at https://t.me/pythontelegrambotgroup. Questions asked there are usually answered more quickly, as there are some 7K members.
